### PR TITLE
[SPARK-21376][YARN] Fix yarn client token expire issue when cleaning the staging files in long running scenario

### PR DESCRIPTION
**Pull request by jerryshao**
## What changes were proposed in this pull request?

This issue happens in long running application with yarn cluster mode, because yarn#client doesn't sync token with AM, so it will always keep the initial token, this token may be expired in the long running scenario, so when yarn#client tries to clean up staging directory after application finished, it will use this expired token and meet token expire issue.

## How was this patch tested?

Manual verification is secure cluster.

### CODE REVIEW

**Comment by jerryshao** [comment]
@tgravescs @vanzin can you please help to review, thanks!

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 1]
```diff
@@ -192,16 +193,29 @@ private[spark] class Client(
...
-    val stagingDirPath = new Path(appStagingBaseDir, getAppStagingDir(appId))
-    try {
-      val preserveFiles = sparkConf.get(PRESERVE_STAGING_FILES)
-      val fs = stagingDirPath.getFileSystem(hadoopConf)
-      if (!preserveFiles && fs.delete(stagingDirPath, true)) {
-        logInfo(s"Deleted staging directory $stagingDirPath")
+    def cleanupStagingDirInternal(): Unit = {
+      val stagingDirPath = new Path(appStagingBaseDir, getAppStagingDir(appId))
+      try {
+        val preserveFiles = sparkConf.get(PRESERVE_STAGING_FILES)
```
Comment (vanzin) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: You could do this check before even trying anything else in this method (just return early).

**Review by vanzin — COMMENTED** [review]
I really wish YARN-2261 was a thing that existed...

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala (hunk 1 above)
Comment (jerryshao) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: Sure, I will change the code.

**Comment by vanzin** [comment]
Merging to master.

**Comment by dongjoon-hyun** [comment]
Hi, All.
Can we have this on `branch-2.2`, too?

**Pull request closed without merging**